3.next - Start adding Response helpers and fix some issues.

markstory opened the pull request:
Add `withLocation()` and `withType()`. These are immutable versions of `type()` and `location()`. Unfortunately this spiraled a bit out of control as I wanted to keep the tests passing. However, I've:

* Fixed getHeaders() not always following PSR7 interface requirements  when headers are set with `header()`.
* Restored behavior where the status code is set to 302 when a location  header is set and the current status is 200.
* Eagerly set the 'Content-Type' header so that charset(), type() and  withType() all result in the content-type header being correctly set.   This also fixes an issue where the content type would not be correctly  set when the response was emitted.

Refs #9636

markstory committed:
Add withLocation()

Add an immutable variant of location().

markstory committed:
Add deprecation notices to a few methods

These methods should be avoided in favour of using immutable methods
instead and PSR7 interface methods.

markstory committed:
Fix a number of header related issues and clean up tests.

Unfortunately this spiraled a bit out of control as I wanted to keep the
tests passing. However, I've:

* Fixed getHeaders() not always following PSR7 interface requirements
  when headers are set with `header()`.
* Restored behavior where the status code is set to 302 when a location
  header is set and the current status is 200.
* Eagerly set the 'Content-Type' header so that charset(), type() and
  withType() all result in the content-type header being correctly set.
  This also fixes an issue where the content type would not be correctly
  set when the response was emitted.

markstory commented on src/Network/Response.php:
```diff
@@ -686,16 +688,36 @@ public function header($header = null, $value = null)
+            $this->headers[$header] = is_array($value) ? array_map('trim', $value) : [trim($value)];
+        }
+
+        return $this->getSimpleHeaders();
```
This fixes the return of `header()` not returning a simple dictionary when headers are set with the PSR7 methods.

markstory commented on src/Network/Response.php:
```diff
@@ -717,12 +741,34 @@ public function location($url = null)
+        $new = $this->withHeader('Location', $url);
+        if ($new->_status === 200) {
+            $new->_status = 302;
+        }
```
This replicates behavior that used to exist in send() which is not used in the PSR stack.

markstory committed:
Fix phpcs errors.

lorenzo commented on src/Network/Response.php:
```diff
@@ -559,9 +563,6 @@ protected function _setCookies()
     {
-        if (in_array($this->_status, [304, 204])) {
-            return;
-        }
```
So what will happen now when sending those status codes?

markstory replied:
Right now a content-type will be emitted. I'm planning on changing the status code methods so that they clear the content-type though. I'm not entirely sure that linking the status code and content-type header is transparent to the end user though.

lorenzo commented on src/Network/Response.php:
```diff
@@ -519,7 +520,10 @@ public function sendHeaders()
+
+        if (in_array($this->_status, [304, 204])) {
+            $this->_setContentType();
+        }
```
This does not seem equivalent to what we had before. The intention was to avoid setting a content type on those 2 cases, but now it is somehow the opposite.

markstory replied:
Derp.

markstory committed:
Fix inverted condition.

markstory commented:
I'll get the remaining tests fixed later today.

markstory committed:
Fix failing tests.

Don't always collapse headers when header() is called. Also update the
tests to include the automatic content-type header setting.

markstory commented:
Anything else that needs to be done on this? I'm going to fix up the issues where `Content-Type` is sent on 302/204 responses in a follow up pull request. Right now the behavior is not consistent.
